yasa/detection.py

```diff
@@ -341,6 +342,108 @@ def _coincidence(x, y):
 
         return coinc_mat
 
+    def compare_channels(self, score="f1", max_distance_sec=0):
+        """
+        Compare detected events across channels.
+        See full documentation in the methods of SpindlesResults and SWResults.
+        """
+        from itertools import product
+
+        assert score in ["f1", "precision", "recall"], f"Invalid scoring metric: {score}"
+
+        # Extract events and channel
+        detected = self.summary()
+        chan = detected["Channel"].unique()
+
+        # Get indices of start in deciseconds, rounding to nearest deciseconds (100 ms).
+        # This is needed for three reasons:
+        # 1. Speed up the for loop
+        # 2. Avoid memory error in yasa.compare_detection
+        # 3. Make sure that max_distance works even when self and other have different sf.
+        # TODO: Only the Start of the event is currently supported. Add more flexibility?
+        detected["Start"] = (detected["Start"] * 10).round().astype(int)
+        max_distance = int(10 * max_distance_sec)
+
+        # Initialize output dataframe / dict
+        scores = pd.DataFrame(index=chan, columns=chan, dtype=float)
+        scores.index.name = "Channel"
+        scores.columns.name = "Channel"
+        pairs = list(product(chan, repeat=2))
+
+        # Loop across pair of channels
+        for c_index, c_col in pairs:
+            idx_chan1 = detected[detected["Channel"] == c_index]["Start"]
+            idx_chan2 = detected[detected["Channel"] == c_col]["Start"]
+            # DANGER: Note how we invert idx_chan2 and idx_chan1 here. This is because
+            # idx_chan1 (the index of the dataframe) should be the ground-truth.
+            res = compare_detection(idx_chan2, idx_chan1, max_distance)
+            scores.loc[c_index, c_col] = res[score]
+
+        return scores
+
+    def compare_detection(self, other, max_distance_sec=0, other_is_groundtruth=True):
+        """
+        Compare detected events between two detection methods, or against a ground-truth scoring.
+        See full documentation in the methods of SpindlesResults and SWResults.
+        """
+        detected = self.summary()
+        if isinstance(other, (SpindlesResults, SWResults, REMResults)):
+            groundtruth = other.summary()
+        elif isinstance(other, pd.DataFrame):
+            assert "Start" in other.columns
```

Comment:
The start of the detected spindle is used as a reference point for comparison. This might not be the best practice to find event matches. For example, Lacourse et al., 2019; LaRocco et al., 2018; and Warby et al., 2014 describe one of their common methods as - a spindle event match is found if the ratio between the intersection and union of the respective spindle intervals is greater than 0.2. 0.2 is only a threshold. What are your thoughts on it?

I think this method of starting with a look around period is fine from a technical standpoint!

Reply:
Thanks @umair-hassan. I think that having the method that you suggest would be a great addition. It reminds me of the [yasa.SpindlesResults.get_coincidence_matrix](https://raphaelvallat.com/yasa/build/html/generated/yasa.SpindlesResults.html#yasa.SpindlesResults.get_coincidence_matrix) method.

That said, I do like the current implementation because it only requires the indices (in seconds from beginning of recording) of specific events (e.g. the start of the spindles). The intersection/union method, if I understand correctly, would require the full boolean mask ([0, 1, 1, 0, ... ] at each sample) of the two detection. I see two potential issues with this:
1. We will need to resample if the two detections do not have the same sampling frequency, which may introduce errors
2. Many detectors (for example ECG heartbeat) out there only return specific time points instead of a full-resolution boolean mask. Therefore, the current implementation of yasa.compare_detection may be applicable to a larger number of use cases.

Still, we could perhaps support both methods and allow users to switch between `method="distance"` (current implementation) or `method="intersection"`.

Reply:
I agree with your points here, so let's just use a method ='distance' here in this case and merge, we can still open another issue for the intersection method implementation for future releases.

Reply:
Awesome, just opened a new issue: https://github.com/raphaelvallat/yasa/issues/88